tensorflow/core/kernels/conv_2d.h

```diff
@@ -189,60 +163,6 @@ struct TransformFilter {
   }
 };
 
-template <typename Device, typename T, typename IndexType>
-struct TransformDepth {
-  void operator()(const Device& d,
-                  typename TTypes<T, 4, IndexType>::ConstTensor in,
-                  const Eigen::DSizes<IndexType, 4>& shuffle,
-                  typename TTypes<T, 4, IndexType>::Tensor out) {
-    Eigen::DSizes<IndexType, 3> merged_dims;
-    Eigen::DSizes<IndexType, 4> expanded_dims;
-    Eigen::DSizes<IndexType, 3> new_shuffle;
-
-    // Merge dimensions that won't be shuffled together to speed things up.
-    if (shuffle[1] == 2 && shuffle[2] == 3) {
-      merged_dims[0] = in.dimension(0);
-      merged_dims[1] = in.dimension(1);
-      merged_dims[2] = in.dimension(2) * in.dimension(3);
-      new_shuffle[0] = shuffle[0];
-      new_shuffle[1] = 2;
-      new_shuffle[2] = shuffle[3];
-      expanded_dims[0] = in.dimension(shuffle[0]);
-      expanded_dims[1] = in.dimension(2);
-      expanded_dims[2] = in.dimension(3);
-      expanded_dims[3] = in.dimension(shuffle[3]);
-    } else if (shuffle[0] == 2 && shuffle[1] == 3) {
-      merged_dims[0] = in.dimension(0);
-      merged_dims[1] = in.dimension(1);
-      merged_dims[2] = in.dimension(2) * in.dimension(3);
-      new_shuffle[0] = 2;
-      new_shuffle[1] = shuffle[2];
-      new_shuffle[2] = shuffle[3];
-      expanded_dims[0] = in.dimension(2);
-      expanded_dims[1] = in.dimension(3);
-      expanded_dims[2] = in.dimension(shuffle[2]);
-      expanded_dims[3] = in.dimension(shuffle[3]);
-    } else if (shuffle[0] == 0 && shuffle[1] == 3 && shuffle[2] == 1 &&
-               shuffle[3] == 2) {
-      merged_dims[0] = in.dimension(0);
-      merged_dims[1] = in.dimension(1) * in.dimension(2);
-      merged_dims[2] = in.dimension(3);
-      new_shuffle[0] = 0;
-      new_shuffle[1] = 2;
-      new_shuffle[2] = 1;
-      expanded_dims[0] = in.dimension(0);
-      expanded_dims[1] = in.dimension(3);
-      expanded_dims[2] = in.dimension(1);
-      expanded_dims[3] = in.dimension(2);
-    } else {
-      assert(false && "unexpected shuffle");
-    }
-
-    out.device(d) =
```

Comment:
Unfortunately, fused_conv2d_bias_activation_op.cc is relying on this statement to define the following template and removing this causes it to fail with the following error. I could not figure out the source of the template usage, so for now, we could leave this and the declaration for float in conv_2d_gpu_float.cu.cc with a TODO.


ld.lld: error: undefined symbol: Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<int, 4> const, Eigen::TensorShufflingOp<Eigen::DSizes<int, 3> const, Eigen::TensorReshapingOp<Eigen::DSizes<int, 3> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, int>, 16, Eigen::MakePointer> const> const> const> const> const, Eigen::GpuDevice, false, false>::run(Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<int, 4> const, Eigen::TensorShufflingOp<Eigen::DSizes<int, 3> const, Eigen::TensorReshapingOp<Eigen::DSizes<int, 3> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, int>, 16, Eigen::MakePointer> const> const> const> const> const&, Eigen::GpuDevice const&)
>>> referenced by fused_conv2d_bias_activation_op.cc
tensorflow::LaunchFusedConv2DBiasActivationOp<Eigen::GpuDevice, float, float, float>::launch(tensorflow::OpKernelContext*, bool, tensorflow::Tensor const&, float, tensorflow::Tensor const&, int, int, Eigen::PaddingType const&, tensorflow::Tensor const&, float, tensorflow::Tensor const&, tensorflow::ActivationMode, tensorflow::TensorFormat, tensorflow::FilterTensorFormat, tensorflow::Tensor*)

Reply:
Thanks for taking a look! How can I locally reproduce this issue? The build is passing for me (and also seems to pass in the CI at least on Linux)

Reply:
Could you try building the following target with GPU? This is one of the failing targets.

third_party/tensorflow/contrib/factorization/examples:mnist

Reply:
It seems there is no such target `third_party/tensorflow/contrib/factorization/examples:mnist` though, at least not in the github clone.

I found a target without third_party and `bazel build --config=cuda -c opt //tensorflow/contrib/factorization/examples:mnist --verbose_failures` passed for me.

Reply:
Yes, that is the target that is failing internally. Which compiler are you using? clang or nvcc? The failure might be specific to clang.

Reply:
I was using nvcc, but clang build also passed for me.
What I did: 
1. `bazel clean` and reconfigure with clang turned on.
2. Apply some patches (otherwise even the master won't build)
3. `bazel build  --config=cuda_clang //tensorflow/contrib/factorization/examples:mnist --verbose_failures`

This is the patch I applied. It's partially from #27348
```diff
diff --git i/third_party/eigen3/gpu_packet_math.patch w/third_party/eigen3/gpu_packet_math.patch
index 3d58cafd03..82102038fc 100644
--- i/third_party/eigen3/gpu_packet_math.patch
+++ w/third_party/eigen3/gpu_packet_math.patch
@@ -4,7 +4,7 @@
    return make_double2(from, from);
  }

-+#if defined(EIGEN_CUDA_ARCH)
++#if __clang__ || defined(EIGEN_CUDA_ARCH)
  namespace {

  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE float bitwise_and(const float& a,
diff --git i/third_party/gpus/crosstool/CROSSTOOL.tpl w/third_party/gpus/crosstool/CROSSTOOL.tpl
index 53d6039d73..a065214ef9 100644
--- i/third_party/gpus/crosstool/CROSSTOOL.tpl
+++ w/third_party/gpus/crosstool/CROSSTOOL.tpl
@@ -91,7 +91,7 @@ toolchain {
         # We need to undef it before redefining it as some distributions now
         # have it enabled by default.
         flag: "-U_FORTIFY_SOURCE"
-        flag: "-D_FORTIFY_SOURCE=1"
+        #flag: "-D_FORTIFY_SOURCE=1"
         flag: "-fstack-protector"
       }
     }
diff --git i/third_party/toolchains/clang6/CROSSTOOL.tpl w/third_party/toolchains/clang6/CROSSTOOL.tpl
index ffba9850bb..440887ab1d 100644
--- i/third_party/toolchains/clang6/CROSSTOOL.tpl
+++ w/third_party/toolchains/clang6/CROSSTOOL.tpl
@@ -70,7 +70,7 @@ toolchain {
   # We need to undef it before redefining it as some distributions now have
   # it enabled by default.
   compiler_flag: "-U_FORTIFY_SOURCE"
-  compiler_flag: "-D_FORTIFY_SOURCE=1"
+  #compiler_flag: "-D_FORTIFY_SOURCE=1"
   compiler_flag: "-fstack-protector"
   linker_flag: "-Wl,-z,relro,-z,now"

diff --git i/third_party/toolchains/preconfig/ubuntu16.04/clang/CROSSTOOL w/third_party/toolchains/preconfig/ubuntu16.04/clang/CROSSTOOL
index 48f82eb35d..8bda0754a4 100755
--- i/third_party/toolchains/preconfig/ubuntu16.04/clang/CROSSTOOL
+++ w/third_party/toolchains/preconfig/ubuntu16.04/clang/CROSSTOOL
@@ -113,7 +113,7 @@ toolchain {
     mode: OPT
     compiler_flag: "-g0"
     compiler_flag: "-O2"
-    compiler_flag: "-D_FORTIFY_SOURCE=1"
+    #compiler_flag: "-D_FORTIFY_SOURCE=1"
     compiler_flag: "-DNDEBUG"
     compiler_flag: "-ffunction-sections"
     compiler_flag: "-fdata-sections"
@@ -144,7 +144,7 @@ toolchain {
       }
       }
     }
-  
+

   feature {
     name: 'fdo_optimize'
```

Reply:
Cool! I’m also aware of this problem and this might help: 
https://bugs.llvm.org/show_bug.cgi?id=41390

Reply:
> Unfortunately, fused_conv2d_bias_activation_op.cc is relying on this statement to define the following template and removing this causes it to fail with the following error. I could not figure out the source of the template usage, so for now, we could leave this and the declaration for float in conv_2d_gpu_float.cu.cc with a TODO.
> 
> ld.lld: error: undefined symbol: Eigen::internal::TensorExecutor<Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<int, 4> const, Eigen::TensorShufflingOp<Eigen::DSizes<int, 3> const, Eigen::TensorReshapingOp<Eigen::DSizes<int, 3> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, int>, 16, Eigen::MakePointer> const> const> const> const> const, Eigen::GpuDevice, false, false>::run(Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<float, 4, 1, int>, 16, Eigen::MakePointer>, Eigen::TensorReshapingOp<Eigen::DSizes<int, 4> const, Eigen::TensorShufflingOp<Eigen::DSizes<int, 3> const, Eigen::TensorReshapingOp<Eigen::DSizes<int, 3> const, Eigen::TensorMap<Eigen::Tensor<float const, 4, 1, int>, 16, Eigen::MakePointer> const> const> const> const> const&, Eigen::GpuDevice const&)
> 
> > > > referenced by fused_conv2d_bias_activation_op.cc
> > > > tensorflow::LaunchFusedConv2DBiasActivationOp<Eigen::GpuDevice, float, float, float>::launch(tensorflow::OpKernelContext*, bool, tensorflow::Tensor const&, float, tensorflow::Tensor const&, int, int, Eigen::PaddingType const&, tensorflow::Tensor const&, float, tensorflow::Tensor const&, tensorflow::ActivationMode, tensorflow::TensorFormat, tensorflow::FilterTensorFormat, tensorflow::Tensor*)

nice

Reply:
ok

Reply:
It is still failing.

What do you think about my earlier proposal of leaving this behind for now and then we can revisit this when I get some time to further investigate it?

Reply:
Sure! Now I kept `TransformDepth` as well as the declaration for float.